File: SolaceApp/Services/FirestoreCircleService.swift
import Foundation
import FirebaseFirestore
import SolaceCore

/// Firebase-backed implementation of `CircleServicing`. Circles themselves
/// are seeded via the console (see `firestore.rules` — create/delete are
/// disallowed from the client); this service only supports browsing,
/// joining/leaving, and posting within a circle.
final class FirestoreCircleService: CircleServicing, @unchecked Sendable {
    private let firestore = Firestore.firestore()

    func observeCircles() -> AsyncStream<[SupportCircle]> {
        AsyncStream { continuation in
            nonisolated(unsafe) let listener = firestore.collection("circles")
                .order(by: "name")
                .addSnapshotListener { snapshot, _ in
                    guard let snapshot else { return }
                    continuation.yield(snapshot.documents.compactMap(Self.circle(from:)))
                }
            continuation.onTermination = { _ in listener.remove() }
        }
    }

    func join(circleID: String, userID: String) async throws {
        try await firestore.collection("circles").document(circleID).setData([
            "memberIDs": FieldValue.arrayUnion([userID])
        ], merge: true)
    }

    func leave(circleID: String, userID: String) async throws {
        try await firestore.collection("circles").document(circleID).setData([
            "memberIDs": FieldValue.arrayRemove([userID])
        ], merge: true)
    }

    func observeMessages(circleID: String) -> AsyncStream<[CircleMessage]> {
        AsyncStream { continuation in
            nonisolated(unsafe) let listener = firestore.collection("circles").document(circleID)
                .collection("messages")
                .order(by: "sentAt")
                .addSnapshotListener { snapshot, _ in
                    guard let snapshot else { return }
                    let messages = snapshot.documents.compactMap { Self.message(from: $0, circleID: circleID) }
                    continuation.yield(messages)
                }
            continuation.onTermination = { _ in listener.remove() }
        }
    }

    func sendMessage(circleID: String, senderID: String, senderName: String, text: String) async throws {
        try await firestore.collection("circles").document(circleID).collection("messages").addDocument(data: [
            "senderID": senderID,
            "senderName": senderName,
            "text": text,
            "sentAt": FieldValue.serverTimestamp()
        ])
    }

    private static func circle(from document: QueryDocumentSnapshot) -> SupportCircle? {
        let data = document.data()
        guard let name = data["name"] as? String else { return nil }
        return SupportCircle(
            id: document.documentID,
            name: name,
            topicDescription: data["topicDescription"] as? String ?? "",
            memberIDs: data["memberIDs"] as? [String] ?? [],
            createdAt: (data["createdAt"] as? Timestamp)?.dateValue() ?? Date()
        )
    }

    private static func message(from document: QueryDocumentSnapshot, circleID: String) -> CircleMessage? {
        let data = document.data()
        guard
            let senderID = data["senderID"] as? String,
            let text = data["text"] as? String
        else { return nil }
        return CircleMessage(
            id: document.documentID,
            circleID: circleID,
            senderID: senderID,
            senderName: data["senderName"] as? String ?? "Member",
            text: text,
            sentAt: (data["sentAt"] as? Timestamp)?.dateValue() ?? Date()
        )
    }
}

File: SolaceAndroid/app/src/main/java/com/jonathanalumasa/solace/service/firebase/FirestoreCircleService.kt
package com.jonathanalumasa.solace.service.firebase

import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.FieldValue
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.SetOptions
import com.jonathanalumasa.solace.model.CircleMessage
import com.jonathanalumasa.solace.model.SupportCircle
import com.jonathanalumasa.solace.service.CircleService
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.tasks.await
import java.util.Date

/**
 * Firebase-backed [CircleService]. Circles themselves are seeded server-side
 * (firestore.rules disallows client create/delete); this service only supports
 * browsing, joining/leaving, and posting within a circle.
 */
class FirestoreCircleService(
    private val firestore: FirebaseFirestore = FirebaseFirestore.getInstance()
) : CircleService {

    override fun observeCircles(): Flow<List<SupportCircle>> =
        firestore.collection("circles")
            .orderBy("name")
            .snapshotListFlow(::circle)

    override suspend fun join(circleID: String, userID: String) {
        firestore.collection("circles").document(circleID)
            .set(mapOf("memberIDs" to FieldValue.arrayUnion(userID)), SetOptions.merge())
            .await()
    }

    override suspend fun leave(circleID: String, userID: String) {
        firestore.collection("circles").document(circleID)
            .set(mapOf("memberIDs" to FieldValue.arrayRemove(userID)), SetOptions.merge())
            .await()
    }

    override fun observeMessages(circleID: String): Flow<List<CircleMessage>> =
        firestore.collection("circles").document(circleID)
            .collection("messages")
            .orderBy("sentAt")
            .snapshotListFlow { message(it, circleID) }

    override suspend fun sendMessage(
        circleID: String,
        senderID: String,
        senderName: String,
        text: String
    ) {
        firestore.collection("circles").document(circleID)
            .collection("messages")
            .add(
                mapOf(
                    "senderID" to senderID,
                    "senderName" to senderName,
                    "text" to text,
                    "sentAt" to FieldValue.serverTimestamp()
                )
            )
            .await()
    }

    private fun circle(document: DocumentSnapshot): SupportCircle? {
        val name = document.getString("name") ?: return null
        @Suppress("UNCHECKED_CAST")
        val memberIDs = document.get("memberIDs") as? List<String> ?: emptyList()
        return SupportCircle(
            id = document.id,
            name = name,
            topicDescription = document.getString("topicDescription") ?: "",
            memberIDs = memberIDs,
            createdAt = document.getDate("createdAt") ?: Date()
        )
    }

    private fun message(document: DocumentSnapshot, circleID: String): CircleMessage? {
        val senderID = document.getString("senderID") ?: return null
        val text = document.getString("text") ?: return null
        return CircleMessage(
            id = document.id,
            circleID = circleID,
            senderID = senderID,
            senderName = document.getString("senderName") ?: "Member",
            text = text,
            sentAt = document.getDate("sentAt") ?: Date()
        )
    }
}
